bamboolean/lexer.py

```diff
@@ -105,65 +107,51 @@ def _integer(self):
             self.next()
         return int(result)
 
+    def skip_and_invoke(self, count, func):
+        for i in range(count):
+            self.next()
+        return func()
+
+    def is_token_equal(self, expected):
+        return expected == reduce(
+            lambda actual, _: actual + str(self.peek()),
+            [self.current_char] * len(expected),
+        )
+
     def get_next_token(self):
         """
-        Lexical anaylzer (tokenizer). Breaks sentence apart into tokens
+        Lexical analyzer (tokenizer). Breaks sentence apart into tokens
         """
         while self.current_char is not None:
             if self.current_char.isspace():
                 self.skip_whitespace()
                 continue
 
-            if re.match('[_a-zA-Z]', self.current_char):
-                return self.id()
-
-            if self._is_quotation_mark(self.current_char):
-                self.next()
-                return self.string()
-
-            if self.current_char.isdigit():
-                return self.number()
-
-            # Relational operators
-
-            if self.current_char is '=' and self.peek() == '=':
-                self.next()
-                self.next()
-                return Token(tok.EQ, '==')
-
-            if self.current_char is '!' and self.peek() == '=':
-                self.next()
-                self.next()
-                return Token(tok.NE, '!=')
-
-            if self.current_char is '<':
-                # LTE, LT
-                self.next()
-                if self.current_char is '=':
-                    self.next()
-                    return Token(tok.LTE, '<=')
-                else:
-                    return Token(tok.LT, '<')
-
-            if self.current_char is '>':
-                # GTE, GT
-                self.next()
-                if self.current_char is '=':
-                    self.next()
-                    return Token(tok.GTE, '>=')
-                else:
-                    return Token(tok.GT, '>')
-
-            # Parentheses
-
-            if self.current_char == '(':
-                self.next()
-                return Token(tok.LPAREN, '(')
-
-            if self.current_char == ')':
-                self.next()
-                return Token(tok.RPAREN, ')')
-
+            regex_map = OrderedDict((
+                (r'("|\')', lambda: self.skip_and_invoke(1, self.string)),
+                (r'[_a-zA-Z]', self.id),
+                (r'\d', self.number),
+            ))
+
+            tokens_map = OrderedDict((
+                ('==', Token(tok.EQ, '==')),
+                ('!=', Token(tok.NE, '!=')),
+                ('<=', Token(tok.LTE, '<=')),
+                ('>=', Token(tok.GTE, '>=')),
+                ('>', Token(tok.GT, '>')),
+                ('<', Token(tok.LT, '<')),
+                ('(', Token(tok.LPAREN, '(')),
+                (')', Token(tok.RPAREN, ')')),
+            ))
```

Comment:
minor: there's no need for regex_map and tokens_map to be under while loop

Reply:
Good point, done.